### PR TITLE
Fixes to Spark date_diff transpilation and Spark operator precedence

### PR DESCRIPTION
**Pull request by RCHowell**
*Description of changes:*

```sql
-- YEAR
floor(months_between(<d2>,<d1>) / 12)

-- MONTH
floor(months_between(<d2>, <d1>)

-- DAY
date_diff(<d2>,<d1>)

-- HOUR
SELECT floor((unix_timestamp(<d2>) - unix_timestamp(<d1>)) / 3600);

-- MINUTE
SELECT floor((unix_timestamp(<d2>) - unix_timestamp(<d1>)) / 60);

-- SECOND
SELECT unix_timestamp(<d2>) - unix_timestamp(<d1>);
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/kotlin/org/partiql/scribe/sql/SqlDialect.kt [hunk 1]
```diff
@@ -69,19 +69,27 @@ abstract class SqlDialect : AstBaseVisitor<SqlBlock, SqlBlock>() {
     // IDENTIFIERS & PATHS
 
     /**
-     * Default behavior is to wrap all SFW queries with parentheses.
+     * Default behavior is to wrap all SFW queries (and binary operators) with parentheses.
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/sql/SqlDialect.kt:72]: nit: based on my understanding, this is for nested binary operators

```suggestion
     * Default behavior is to wrap all SFW queries (and nested binary operators) with parentheses.
```

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 2]
```diff
@@ -119,30 +115,83 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
-            else -> error("Unexpected datetime part `$part`")
+    override fun dateDiff(part: DatetimeField, args: SqlArgs): Expr = when (part) {
+        DatetimeField.YEAR -> {
+            val d1 = args[0].expr
+            val d2 = args[1].expr
+            val call = exprCall(
+                function = id("months_between"),
+                args = listOf(d2, d1)
+            )
+            log.info("PartiQL `date_diff(year, <date_1>, <date_2>)` was replaced by Spark `floor(months_between(<date_2>, <date_1>) / 12).")
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:134]: I'm not sure if `floor` is what we want for `date_diff` (months, years, hour, and minute). Consider if first datetime is earlier than the second datetime, the `months_between`/`diff` will return a negative value.

E.g.: date_diff between two timestamps that are 1 minute and 1 second apart

```
PartiQL> DATE_DIFF(minute, `1997-02-28T10:30:01Z`, `1997-02-28T10:29:00Z`);
==='
-1
```

spark will output something different

```
spark-sql (default)> SELECT floor((unix_timestamp('1997-02-28 10:29:00') - unix_timestamp('1997-02-28 10:30:01')) / 60);
-2
```

Probably what we need is some kind of decimal truncation or rounding to 0?

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 3]
```diff
@@ -119,30 +115,83 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
+            val d1 = unixTimestamp(args[0].expr)
+            val d2 = unixTimestamp(args[1].expr)
+            val d = diff(d2, d1)
+            floor(div(d, 60))
+        }
+        DatetimeField.SECOND -> {
+            val d1 = unixTimestamp(args[0].expr)
+            val d2 = unixTimestamp(args[1].expr)
+            diff(d2, d1)
+        }
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:172]: nit: could add `log.info` on these translations as well

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt (hunk 2 above)
Comment (RCHowell) [reply src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:134]: Good call. `CAST(.. AS INT)` will truncate.

Comment (RCHowell) [reply src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:134]:
```sql
spark-sql (default)> SELECT CAST((`unix_timestamp`(TIMESTAMP'1997-02-28T10:29:00Z') - `unix_timestamp`(TIMESTAMP'1997-02-28T10:30:01Z')) / 60 AS BIGINT);
-1
```

**Review by alancai98 — APPROVED** [review]

**Pull request merged**